sys/usb/cdc/cdc_acm.c

```diff
@@ -0,0 +1,252 @@
+}
+
+static size_t _gen_acm_descriptor(plumbum_t *plumbum, void *arg)
+{
+    (void)arg;
+    usb_desc_acm_t acm;
+    /* functional cdc acm descriptor */
+    acm.length = sizeof(usb_desc_acm_t);
+    acm.type = USB_TYPE_DESCRIPTOR_CDC;
+    acm.subtype = 0x02;
+    acm.capabalities = 0x00;
+    plumbum_put_bytes(plumbum, (uint8_t*)&acm, sizeof(acm));
+    return sizeof(usb_desc_acm_t);   
+}
+
+static size_t _acm_size(plumbum_t *plumbum, void *arg)
+{
+    (void)plumbum;
+    (void)arg;
+    return sizeof(usb_desc_acm_t);
+}
+
+static size_t _gen_cdc_descriptor(plumbum_t *plumbum, void *arg)
+{
+    (void)arg;
+    usb_desc_cdc_t cdc;
+    /* functional cdc descriptor */
+    cdc.length = sizeof(usb_desc_cdc_t);
+    cdc.bcd_hid = 0x0110;
+    cdc.type = USB_TYPE_DESCRIPTOR_CDC;
+    cdc.subtype = 0x00;
+    plumbum_put_bytes(plumbum, (uint8_t*)&cdc, sizeof(cdc));
+    return sizeof(usb_desc_cdc_t);
+}
+
+static size_t _cdc_size(plumbum_t *plumbum, void *arg)
+{
+    (void)plumbum;
+    (void)arg;
+    return sizeof(usb_desc_cdc_t);
+}
+
+static plumbum_cdc_device_t handler;
+
+int cdc_init(plumbum_t *plumbum)
+{
+    memset(&handler, 0 , sizeof(plumbum_cdc_device_t));
+    handler.handler_ctrl.driver = &cdc_driver;
+    plumbum_register_event_handler(plumbum, (plumbum_handler_t*)&handler);
+    return 0;
+}
+
+static int _init(plumbum_t *plumbum, plumbum_handler_t *handler)
+{
+    plumbum_cdc_device_t *cdc = (plumbum_cdc_device_t*)handler;
+
+    cdc->call_mngt_hdr.next = NULL;
+    cdc->call_mngt_hdr.gen_hdr = _gen_call_mngt_descriptor;
+    cdc->call_mngt_hdr.hdr_len = _call_mngt_size;
+    cdc->call_mngt_hdr.arg = cdc;
+
+    cdc->union_hdr.next = &cdc->call_mngt_hdr;
+    cdc->union_hdr.gen_hdr = _gen_union_descriptor;
+    cdc->union_hdr.hdr_len = _union_size;
+    cdc->union_hdr.arg = cdc;
+
+    cdc->acm_hdr.next = &cdc->union_hdr;
+    cdc->acm_hdr.gen_hdr = _gen_acm_descriptor;
+    cdc->acm_hdr.hdr_len = _acm_size;
+    cdc->acm_hdr.arg = cdc;
+
+    cdc->cdc_hdr.next = &cdc->acm_hdr;
+    cdc->cdc_hdr.gen_hdr = _gen_cdc_descriptor;
+    cdc->cdc_hdr.hdr_len = _cdc_size;
+    cdc->cdc_hdr.arg = cdc;
+
+    /* Instantiate interfaces */
+    memset(&cdc->iface_ctrl, 0, sizeof(plumbum_interface_t));
+    memset(&cdc->iface_data, 0, sizeof(plumbum_interface_t));
+    /* Configure Interface 0 as control interface */
+    cdc->iface_ctrl.class = USB_CLASS_CDC_CONTROL ;
+    cdc->iface_ctrl.subclass = USB_CDC_SUBCLASS_ACM;
+    cdc->iface_ctrl.protocol = USB_CDC_PROTOCOL_NONE;
+    cdc->iface_ctrl.hdr_gen = &cdc->cdc_hdr;
+    cdc->iface_ctrl.handler = handler;
+    cdc->iface_ctrl.idx = 0;
+    /* Configure second interface to handle data endpoint */
+    cdc->iface_data.class = USB_CLASS_CDC_DATA ;
+    cdc->iface_data.subclass = USB_CDC_SUBCLASS_NONE;
+    cdc->iface_data.protocol = USB_CDC_PROTOCOL_NONE;
+    cdc->iface_data.hdr_gen = NULL;
+    cdc->iface_data.handler = handler;
+    cdc->iface_data.idx = 1;
```

Comment:
@bergzand I thought a bit about this. I think we need another way to replace this. (Maybe let `usbus_add_interface` function handle it automaticaly)
Otherwise, we will have issue when developers will declare several interfaces.

Reply:
I am only talking about the  `cdc->iface_data.idx = 1;` here.